jormungandr/src/network/p2p/topology.rs

```diff
@@ -2,171 +2,204 @@
+        let mut inner = self.lock.write().await;
+        let mut gossips = if let Some(recipient) = recipient {
+            inner.topology.gossips_for(recipient)
+        } else {
+            Vec::new()
+        };
+        // If the recipient is not already in the topology
+        // or was not specified poldercast will not return anything.
+        // Let's broadcast out profile anyway
+        if gossips.is_empty() {
+            gossips.push(inner.topology.self_profile().gossip().clone());
+        }
+        gossips.retain(|g| g.address() != *LOCAL_ADDR);
+        Gossips::from(gossips)
+    }
+
+    #[instrument(skip(self, gossips), level = "debug")]
+    pub async fn accept_gossips(&self, gossips: Gossips) {
+        let mut inner = self.lock.write().await;
+        let gossips = <Vec<poldercast::Gossip>>::from(gossips);
+        for gossip in gossips {
+            let peer = Profile::from_gossip(gossip);
+            tracing::trace!(node = %peer.address(), "received peer from gossip");
+            inner.topology.add_peer(peer);
         }
-    }
-
-    pub async fn initiate_gossips(&self, with: Address) -> Gossips {
-        let mut topology = self.lock.write().await;
-        topology.initiate_gossips(with).into()
-    }
-
-    pub async fn accept_gossips(&self, from: Address, gossips: Gossips) {
-        let mut topology = self.lock.write().await;
-        topology.accept_gossips(from, gossips.into())
-    }
-
-    pub async fn exchange_gossips(&mut self, with: Address, gossips: Gossips) -> Gossips {
-        let mut topology = self.lock.write().await;
-        topology.exchange_gossips(with, gossips.into()).into()
-    }
 
-    pub async fn node(&self) -> NodeProfile {
-        let topology = self.lock.read().await;
-        topology.profile().clone()
+        // nodes lifted from quarantine will be considered again in the next update
+        let lifted = inner.quarantine.lift_from_quarantine();
+        for node in lifted {
+            // It may happen that a node is evicted from the dirty pool
+            // in poldercast and then re-enters the topology in the 'pool'
+            // pool, all while we hold the node in quarantine.
+            // If that happens we should not promote it anymore.
+            let is_dirty = inner.topology.peers().dirty().contains(&node.id);
+            if is_dirty {
+                tracing::debug!(node = %node.address, "lifting node from quarantine");
+                inner.topology.promote_peer(&node.id);
+            } else {
+                tracing::debug!(node = %node.address, "node from quarantine have left the dirty pool. skipping it");
+            }
+        }
     }
 
+    #[instrument(skip(self), level = "debug")]
     pub async fn force_reset_layers(&self) {
-        let mut topology = self.lock.write().await;
-        topology.force_reset_layers()
+        tracing::warn!("resetting layers is not supported in this poldercast version");
     }
```

Comment:
Hmm, maybe just go to the now useless periodic task that calls it and cut it out?

Reply:
Should we remove that option from the config file as well?